### PR TITLE
Add check of switch links and cpu load

### PR DESCRIPTION
**Pull request by TuxPowered42**

### CODE REVIEW

File: src/switch.py [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+from argparse import ArgumentParser
+from pysnmp import proto, error
+from pysnmp.entity.rfc3413.oneliner import cmdgen
+from pysnmp.entity.rfc3413.oneliner.cmdgen import (
+        CommunityData,
+        UsmUserData,
+        usmHMACSHAAuthProtocol,
+        usmAesCfb128Protocol,
+        usmDESPrivProtocol,
```
Comment (hasegeli) [inline src/switch.py]: Too much indented

File: src/switch.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
+#!/usr/bin/env python3
+#
+# igmonplugins - Switch checks:
+#                - cpu usage
+#                - port state
+#
+# Copyright (c) 2017, InnoGames GmbH
+#
```
Comment (hasegeli) [inline src/switch.py]: Could you please add a license?  We don't have a global one on this project?

File: src/switch.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
+#!/usr/bin/env python3
```
Comment (hasegeli) [inline src/switch.py]: Wouldn't it work on Python 2?  If so, just `python` is a more portable choice.

File: src/switch.py [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+
+    print(exit_msg)
+    sys.exit(exit_code)
+
+
+def parse_args():
+    parser = ArgumentParser()
+    parser.add_argument(
+            '-H', dest='switch', required=True,
+            type=str, help='Hostname of a switch'
```
Comment (hasegeli) [inline src/switch.py]: Too much indented

File: src/switch.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        exit_msg = 'Unknown check type {} requested'.format(args.check)
+
+    print(exit_msg)
+    sys.exit(exit_code)
+
+
+def parse_args():
+    parser = ArgumentParser()
+    parser.add_argument(
+            '-H', dest='switch', required=True,
```
Comment (hasegeli) [inline src/switch.py]: I don't like `required` optional arguments.  Wouldn't `localhost` be a reasonable default.

File: src/switch.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+            type=str, help='Hostname of a switch'
+    )
+    parser.add_argument(
+        '--check', type=str, help='Check to perform',
+        choices=['cpu', 'links'],
+        default='links',
+    )
+    parser.add_argument(
+        '-v', dest='verbose', action='store_true',
+        help='Verbose output - show OK ports',
```
Comment (hasegeli) [inline src/switch.py]: Some arguments are long some are short.  I prefer to provide long ones only on such scripts intended for non-interactive use, even though they are less standard.

File: src/switch.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    """ Get a single value from SNMP """
+
+    errorIndication, errorStatus, errorIndex, varBinds = cmd_gen.getCmd(
+        snmp['auth_data'],
+        snmp['transport_target'],
+        OID,
+    )
+    if errorIndication:
+        raise SwitchException('Unable to get SNMP value: {}'.
+                              format(errorIndication))
```
Comment (hasegeli) [inline src/switch.py]: Dot before the method please

File: src/switch.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+    return ret
+
+
+def convert_snmp_type(varBinds):
+    """ Convert SNMP data types to something more convenient: int or str """
+    val = varBinds[0][1]
+    if type(val) in [
+        proto.rfc1902.Integer,
+        proto.rfc1902.Counter32,
+        proto.rfc1902.Counter64,
```
Comment (hasegeli) [inline src/switch.py]: Aren't they subclass of each other?  Maybe what you need is `isinstance()`.

File: src/switch.py [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+
+    if 'PowerConnect' in model:
+        return 'powerconnect'
+    elif 'ProCurve' in model:
+        return 'procurve'
+    elif 'ExtremeXOS' in model:
+        return 'extreme'
+
+    print('Unknown switch model', file=sys.stderr)
+    return None
```
Comment (hasegeli) [inline src/switch.py]: Why not raising an `Exception`?

File: src/switch.py [hunk 10]
```diff
@@ -0,0 +1,374 @@
...
+            port_names[port_index].startswith('Vl') or
+            port_names[port_index].startswith('CPU') or
+            port_names[port_index].startswith('Po')
+        )):
+            continue
+
+        if (model == 'procurve' and (
+            port_names[port_index].startswith('VLAN') or
+            port_names[port_index].startswith('Trk')
+        )):
```
Comment (hasegeli) [inline src/switch.py]: One level of `if`/`elif`s for `model`s and `if`s inside it would be more readable.

File: src/switch.py [hunk 11]
```diff
@@ -0,0 +1,374 @@
...
+            continue
+
+        local_exit = 1
+        msg = (
+            'WARNING: Unhandled but bad status (admin: {}, oper:{})!'.
+            format(port_admin_states[port_index], port_oper_states[port_index])
```
Comment (hasegeli) [inline src/switch.py]: Dot before the method please

File: src/switch.py [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+        )
+
+        # The only OK cases are:
+        # Port is named, enabled and up.
+        if (
+            port_aliases[port_index].strip() != '' and
```
Comment (hasegeli) [inline src/switch.py]: It is Pythonic to rely on empty string being `False`.

File: src/switch.py [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+        ):
+            local_exit = 2
+            msg = 'CRITICAL: Unnamed port is up!'
+
+        if (
+            port_aliases[port_index] != '' and
+            port_oper_states[port_index] == 2
+        ):
+            local_exit = 2
+            msg = 'CRITICAL: Named port is down!'
```
Comment (hasegeli) [inline src/switch.py]: I think you can make this also more readable with multiple levels of `if`s.

File: src/switch.py [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+
+        if local_exit > 0 or args.verbose:
+            outmsg += (
+                '{} "{}": {}\n'.
+                format(port_names[port_index], port_aliases[port_index], msg)
```
Comment (hasegeli) [inline src/switch.py]: Dot before the method please

File: src/switch.py [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+    elif 'ProCurve' in model:
+        return 'procurve'
+    elif 'ExtremeXOS' in model:
+        return 'extreme'
+
+    print('Unknown switch model', file=sys.stderr)
+    return None
+
+
+def get_laggs(snmp, model):
```
Comment (hasegeli) [inline src/switch.py]: This is not used.

File: src/switch.py [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+        return {
+            x: y for x, y
+            in get_snmp_table(snmp, LAGG_OIDS[model]).items()
+            if y != 0
+        }
+
+    return None
+
+
+def standarize_portname(port_name):
```
Comment (hasegeli) [inline src/switch.py]: This neither

File: src/switch.py (hunk 3 above)
Comment (TuxPowered42) [inline src/switch.py]: I have no idea if it works with Python2 and I refuse to check it.

File: src/switch.py (hunk 2 above)
Comment (TuxPowered42) [inline src/switch.py]: Added.

File: src/switch.py (hunk 1 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 4 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 7 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 15 above)
Comment (TuxPowered42) [inline src/switch.py]: Right, it came from the previous script. I will remove it.

File: src/switch.py (hunk 16 above)
Comment (TuxPowered42) [inline src/switch.py]: This too.

File: src/switch.py (hunk 14 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 13 above)
Comment (TuxPowered42) [inline src/switch.py]: That would be less readable IMO.

File: src/switch.py (hunk 11 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 10 above)
Comment (TuxPowered42) [inline src/switch.py]: True, fixed.

File: src/switch.py (hunk 12 above)
Comment (TuxPowered42) [inline src/switch.py]: Maybe pythonic, but quite unreadable code. I want to know what the string contains by first look on it, not by deep knowledge of language internals.

File: src/switch.py (hunk 6 above)
Comment (TuxPowered42) [inline src/switch.py]: I dunno. I've seen `-v` in way too many programs as verbose, to consider the long version. The other short parameters are de facto Nagios plugin standard. I really don't feel like forcing long or short on all of them. That would not really make the world a better place and if anybody is confused, there is `-h` for them.

File: src/switch.py (hunk 8 above)
Comment (TuxPowered42) [inline src/switch.py]: They don't seem to be.

File: src/switch.py (hunk 9 above)
Comment (TuxPowered42) [inline src/switch.py]: Fixed.

File: src/switch.py (hunk 8 above)
Comment (hasegeli) [inline src/switch.py]: They are all subclass of `pyasn1.type.univ.Integer` whatever that is.  It is fine to leave it like this.

**Pull request merged**